src/utils/typedData.ts

```diff
@@ -107,130 +152,183 @@ function getMerkleTreeType(types: TypedData['types'], ctx: Context) {
+
   return newTypes
     .map((dependency) => {
-      return `${dependency}(${types[dependency].map((t) => `${t.name}:${t.type}`)})`;
+      const dependencyElements = types[dependency].map((t) => {
+        const targetType = t.type === 'enum' ? (t as StarkNetEnumType).contains : t.type;
+        // parentheses handling for enum variant types
+        const typeString = targetType.match(/^\(.*\)$/)
+          ? `(${targetType
+              .slice(1, -1)
+              .split(',')
+              .map((e) => (e ? esc(e.trim()) : e))
+              .join(',')})`
+          : esc(targetType);
+        return `${esc(t.name)}:${typeString}`;
+      });
+      return `${esc(dependency)}(${dependencyElements})`;
     })
     .join('');
-};
+}
 
 /**
  * Get a type string as hash.
  */
-export const getTypeHash = (types: TypedData['types'], type: string): string => {
-  return getSelectorFromName(encodeType(types, type));
-};
+export function getTypeHash(
+  types: TypedData['types'],
+  type: string,
+  revision: Revision = Revision.Legacy
+): string {
+  return getSelectorFromName(encodeType(types, type, revision));
+}
 
 /**
  * Encodes a single value to an ABI serialisable string, number or Buffer. Returns the data as tuple, which consists of
  * an array of ABI compatible types, and an array of corresponding values.
  */
-export const encodeValue = (
+export function encodeValue(
   types: TypedData['types'],
   type: string,
   data: unknown,
-  ctx: Context = {}
-): [string, string] => {
+  ctx: Context = {},
+  revision: Revision = Revision.Legacy
+): [string, string] {
   if (types[type]) {
-    // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    return [type, getStructHash(types, type, data as Record<string, unknown>)];
+    return [type, getStructHash(types, type, data as TypedData['message'], revision)];
   }
 
-  if (
-    Object.keys(types)
-      .map((x) => `${x}*`)
-      .includes(type)
-  ) {
-    const structHashes: string[] = (data as unknown[]).map((struct) => {
-      // eslint-disable-next-line @typescript-eslint/no-use-before-define
-      return getStructHash(types, type.slice(0, -1), struct as Record<string, unknown>);
-    });
-    return [type, computeHashOnElements(structHashes)];
+  if (presetTypes[type]) {
+    return [type, getStructHash(presetTypes, type, data as TypedData['message'], revision)];
   }
 
-  if (type === 'merkletree') {
-    const merkleTreeType = getMerkleTreeType(types, ctx);
-    const structHashes: string[] = (data as unknown[]).map((struct) => {
-      // eslint-disable-next-line @typescript-eslint/no-use-before-define
-      return encodeValue(types, merkleTreeType, struct as Record<string, unknown>)[1];
-    });
-    const { root } = new MerkleTree(structHashes as string[]);
-    return ['felt', root];
-  }
+  if (type.endsWith('*')) {
+    const isStructArray = Object.keys(types)
+      .map((x) => `${x}*`)
+      .includes(type);
```

Comment:
not sure but looks like it's taking presets into account

Reply:
Presets should be taken into account through the mapped encodeValue calls.

Reply:
can you double check that?
I tried `TokenAmount*` and `isStructArray` was set to false instead of true

Reply:
It was getting taken into account by the recursive `encodeValue` call. I noticed the code could be simplified so that both custom and preset types rely on it so it should be clearer now.